nexus/src/db/datastore/silo_group.rs

```diff
@@ -94,51 +79,52 @@ impl DataStore {
+    /// where a crash half way through causes the resulting group memberships to
+    /// be incorrect.
+    pub async fn silo_group_membership_replace_for_user(
         &self,
         opctx: &OpContext,
         silo_user_id: Uuid,
-    ) -> ListResultVec<SiloGroupMembership> {
-        use db::schema::silo_group_membership::dsl;
-        dsl::silo_group_membership
-            .filter(dsl::silo_user_id.eq(silo_user_id))
-            .select(SiloGroupMembership::as_returning())
-            .get_results_async(self.pool_authorized(opctx).await?)
+        silo_group_ids: Vec<Uuid>,
+    ) -> UpdateResult<()> {
+        self.pool_authorized(opctx)
+            .await?
+            .transaction(move |conn| {
+                use db::schema::silo_group_membership::dsl;
+
+                // Delete existing memberships for user
+                diesel::delete(dsl::silo_group_membership)
+                    .filter(dsl::silo_user_id.eq(silo_user_id))
+                    .execute(conn)?;
+
+                // Create new memberships for user
+                let silo_group_memberships: Vec<
+                    db::model::SiloGroupMembership,
+                > = silo_group_ids
+                    .iter()
+                    .map(|group_id| db::model::SiloGroupMembership {
+                        silo_group_id: *group_id,
+                        silo_user_id: silo_user_id,
+                    })
+                    .collect();
+
+                diesel::insert_into(dsl::silo_group_membership)
+                    .values(silo_group_memberships)
+                    .execute(conn)?;
+
+                Ok(())
+            })
             .await
-            .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
-    }
-
-    pub async fn silo_group_membership_delete(
-        &self,
-        opctx: &OpContext,
-        authz_silo: &authz::Silo,
-        silo_user_id: Uuid,
-        silo_group_id: Uuid,
-    ) -> DeleteResult {
-        opctx.authorize(authz::Action::ListChildren, authz_silo).await?;
-
-        let (_authz_silo_group, ..) = LookupPath::new(opctx, &self)
-            .silo_group_id(silo_group_id)
-            .fetch_for(authz::Action::Modify)
-            .await?;
-
-        use db::schema::silo_group_membership::dsl;
-        diesel::delete(dsl::silo_group_membership)
-            .filter(dsl::silo_user_id.eq(silo_user_id))
-            .filter(dsl::silo_group_id.eq(silo_group_id))
-            .execute_async(self.pool_authorized(opctx).await?)
-            .await
-            .map_err(|e| {
-                Error::internal_error(&format!(
-                    "error deleting group membership of group {} for user {}: {:?}",
-                    silo_group_id,
-                    silo_user_id,
-                    e,
-                ))
-            })?;
-        Ok(())
+            .map_err(|e: TransactionError<PoolError>| {
+                Error::internal_error(&format!("Transaction error: {}", e))
```

Comment:
This needs to be some form of `ErrorHandler` (e.g., so we don't produce a 500 when the database is just down).

Reply:
done in 34f49bc7